src/cgi-bin/vnstatxml-1.6/man/vnstat.1
.TH VNSTAT 1 "JANUARY 2008" "version 1.6" "User Manuals"
.SH NAME
vnStat \- a console-based network traffic monitor
.SH SYNOPSIS
.B vnstat
[
.B \-Ddhlmqrstuvw?
] [
.B \-i
.I interface
] [
.B \-tr
.I time
] [
.B \-\-cleartop
] [
.B \-\-config
.I file
] [
.B \-\-days
] [
.B \-\-debug
] [
.B \-\-disable
] [
.B \-\-dumpdb
] [
.B \-\-enable
] [
.B \-\-help
] [
.B \-\-hours
] [
.B \-\-iface
.I interface
] [
.B \-\-live
] [
.B \-\-longhelp
] [
.B \-\-months
] [
.B \-\-nick
.I nickname
] [
.B \-\-query
] [
.B \-\-rebuildtotal
] [
.B \-\-reset
] [
.B \-\-short
] [
.B \-\-showconfig
] [
.B \-\-testkernel
] [
.B \-\-top10
] [
.B \-\-traffic
.I time
] [
.B \-\-update
] [
.B \-\-version
] [
.B \-\-weeks
] [
.B \-\-dumpxml
]
.SH DESCRIPTION
.B vnStat
is a console-based network traffic monitor. It keeps a log of hourly,
daily and monthly network traffic for the selected interface(s). However,
it isn't a packet sniffer. The traffic information is analyzed from the
.BR proc (5)
filesystem. That way vnStat can be used even without root permissions.
.SH OPTIONS
.TP
.BI "-d, --days"
Show traffic for days.
.TP
.BI "-h, --hours"
Show traffic for the last 24 hours.
.TP
.BI "-m, --months"
Show traffic for months.
.TP
.BI "-s, --short"
Use short output mode. This mode is also used if more than one
database is available.
.TP
.BI "-t, --top10"
Show all time top10 traffic days.
.TP
.BI "-w, --weeks"
Show traffic for 7 days, current and previous week.
.TP
.BI "-tr " time
Calculate how much traffic goes through the selected interface during
the given
.I time 
seconds. The 
.I time
will be 5 seconds if a number parameter isn't included.
.TP
.BI "-l, --live"
Display current transfer rate for the selected interface in real time
until interrupted. Statistics will be shown after interruption if runtime
was more than 10 seconds.
.TP
.BI "-i, --iface " interface
Select one specific
.I interface
and apply actions to only it.
.TP
.BI "-q, --query"
Force database query mode.
.TP
.BI "-u, --update"
Update all enabled databases or only the one specified with
.B -i
parameter.
.TP
.BI "-r, --reset"
Reset the internal counters in the database for the selected
interface. Use this if the interface goes down and back up,
otherwise that interface will get some extra traffic to its database.
.TP
.BI "--sync"
Synchronize internal counters in the database with interface
counters for the selected interface. Use this if the system is
rebooted but interface counters aren't reseted. Such can occur
when suspend to ram/disk is used.
.TP
.BI "--enable, --disable"
Enable or disable updates for selected interface. Useful for
interfaces that aren't always available, like ppp0. If the interface
goes down it should be disabled in order to avoid errors. Add something
like
.B "vnstat -r --disable -i ppp0"
to the script that's executed when
the interface goes down and
.B "vnstat --enable -i ppp0"
to the up script.
.TP
.BI "-v, --version"
Show current version.
.TP
.BI "--cleartop"
Remove all top10 entries.
.TP
.BI "-?, --help"
Show a command summary.
.TP
.BI "--longhelp"
Show complete options list.
.TP
.BI "--nick " nickname
Set the selected interfaces
.I nickname
as an alias the will be displayed in queries. Usage of
.B -u
is required to save the change.
.TP
.BI "--config " file
Use
.I file
as config file instead of using normal config file search function.
.TP
.BI "--rebuildtotal"
Reset the total traffic counters and recount those using recorded months.
.TP
.BI "--testkernel"
Test if the kernel boot time information always stays the same like it should or
if it's shifting.
.TP
.BI "-D, --debug"
Show additional debug output.
.TP
.BI "--dumpdb"
Instead of showing the database with a formated output, this output will
dump the whole database in a format that should be easy to parse with most
script languages. Use this for example with PHP, Perl or Python to make a
custom  webpage. The dump uses ; as field delimeter.

  active;1                        activity status
  interface;eth0                  name for the interface
  nick;inet                       nick (if given)
  created;1023895272              creation date in Unix time
  updated;1065467100              when the database was updated
  totalrx;569605                  all time total received MB
  totaltx;2023708                 all time total transmitted MB
  currx;621673719                 latest rx value in /proc
  curtx;981730184                 latest tx value in /proc
  totalrxk;644                    total rx kB counter
  totaltxk;494                    total tx kB counter
  btime;1059414541                system boot time in Unix time

Then follows 30 lines like the following

  d;0;1078696800;559;7433;68;557;1

where d = days, 0 = day number in database (0 is today), 1077314401 date in
Unix time, 559 = rx MB, 7433 = tx MB, 68 = rx kB, 557 = tx kB and 1 tells that
vnStat has filled this value and it is in use.

  m;0;1078092000;48649;139704;527;252;1   (x12)
  t;0;1078351200;5979;47155;362;525;1     (x10)
  h;0;1078699800;118265;516545            (x24)

m = months, t = top10 and h = hours, all other fields are in the same order as in days
except hours that doesn't have a separate kB value. For hours the forth and fifth fields
have values in kB.
.TP
.BI "--dumpxml"
This option will give you a choice to dump the relative data in XML format. it can be 
used with " -i ethN [-t|-d|-m|-h|no option] ", for example,

  $ vnstat --dumpxml -i eth1
  <?xml version='1.0' encoding='UTF-8' standalone='yes'?>
  <traffic id='content' p='summary' colnum='0'>
  <us><u id='0' sym=' KB' val='1'/><u id='1' sym=' MB' val='1024'/><u id='2' sym=' GB' val='1048576'/></us>
  <r f1='08/04/06-->08/07/04' x=''><f><s>68.7722</s><u>2</u></f><f><s>4.66</s><u>2</u></f><f><s>73.44</s><u>2</u></f></r>
  <r f1='this month: 08/07' x=''><f><s>293.9648</s><u>1</u></f><f><s>26.58</s><u>1</u></f><f><s>320.54</s><u>1</u></f></r>
  <r f1='this week: 27' x=''><f><s>346.7510</s><u>1</u></f><f><s>35.61</s><u>1</u></f><f><s>1.05</s><u>2</u></f></r>
  <r f1='today: 08/07/04' x=''><f><s>8.9844</s><u>1</u></f><f><s>649.00</s><u>0</u></f><f><s>9.62</s><u>1</u></f></r>
  <r f1='current hour: 12:00' x=''><f><s>13.0000</s><u>0</u></f><f><s>5.00</s><u>0</u></f><f><s>18.00</s><u>0</u></f></r>
  </traffic>

  $ vnstat --dumpxml -i eth1 -m
  <?xml version='1.0' encoding='UTF-8' standalone='yes'?>
  <traffic id='content' p='month' colnum='12'>
  <us><u id='0' sym=' KB' val='1'/><u id='1' sym=' MB' val='1024'/><u id='2' sym=' GB' val='1048576'/></us>
  <r f1='08-04' x='08-04'><f><s>12.2272</s><u>2</u></f><f><s>738.23</s><u>1</u></f><f><s>12.95</s><u>2</u></f></r>
  <r f1='08-05' x='08-05'><f><s>18.9075</s><u>2</u></f><f><s>2.94</s><u>2</u></f><f><s>21.85</s><u>2</u></f></r>
  <r f1='08-06' x='08-06'><f><s>37.3505</s><u>2</u></f><f><s>1002.33</s><u>1</u></f><f><s>38.33</s><u>2</u></f></r>
  <r f1='08-07' x='08-07'><f><s>293.9648</s><u>1</u></f><f><s>26.58</s><u>1</u></f><f><s>320.54</s><u>1</u></f></r>
  <mf><s>37.35</s><u>2</u></mf>
  </traffic>

  Acronym List of the above XML output:

    p: "page" type, indicates the output is for summary|month|top10|day|hour
    colnum: the number of column in the X axis direction when drawing SVG graph
    us: units
    u: unit
    s: size
    r: row
    f: field
    f1: the first field, the time info (day|hour|month|week|total)
    x: the value of the X axis
    mf: the max number of all the Received and Transmited fields

  Description of the XML result:

   Line            Description
   ----------------------------------------------------------------------------------------
   <?xml...>       the XML header

   <traffic ....>  the XML data header, "p" indicates the info type, "colnum"
                   indicates the width of the X axis when drawing a SVG graph

   <us...>         define the unit information, they are defined as (serial, symbol, value)

   <r...>          store the traffic of every time unit
                   (day|hour|month|week|total) every row include four field, the first field is
                   the time info, the following three fields are respectively Received,
                   Transmited, and Total traffic. the traffic are described as size(s) and
                   unit(u). the unit is defined in the <us...> line.  so,
                   <f><s>12.2272</s><u>2</u></f> means 12.2272GB or 12.2272*1048576KB

   <mf..>          the max traffic of every Received and Transmited Traffic in the above rows.
   -------------------------------------------------------------------------------------------

This option is added by Wu Zhangjin, it is designed for vnStatSVG(http://vnstatsvg.sourceforge.net).

.SH FILES
.TP
.I /var/lib/vnstat/
This directory contains all databases the program uses. Files are
named according to the monitored interfaces.
.TP
.I /etc/vnstat.conf
Config file that will be used unless
.I $HOME/.vnstatrc
exists.
.SH EXAMPLES
.BI "vnstat -u -i" 
.I interface
forces a database update for
.I interface
or creates the database if it doesn't exist. This is usually the
first command used after a fresh install.
.PP
.BI "vnstat -u -i"
.I interface
.BI "--nick"
.I nick
gives
.I interface
the nickname
.I "nick"
and that information will be later included with queries.
.PP
.BI "vnstat -u -r --disable -i
.I interface
resets the internal counters of
.I interface
and disables it from being updated before enabled again with the
.BI "--enable"
parameter. This feature is especially useful for interfaces like ppp0
that aren't always active.
.SH RESTRICTIONS
Updates needs to be executed at least as often as it is possible for the interface
to generate enough traffic to wrap the kernel interface traffic counter. Otherwise
it is possible that some traffic won't be seen. This isn't an issue for 64 bit kernels
but at least one update every hour is always required in order to provide proper input.
With 32 bit kernels the maximum time between two updates depends on how fast the
interface can transfer 4 GB. Calculated theoretical times are:
.PP
    10 Mbit:  54 minutes
   100 Mbit:   5 minutes
  1000 Mbit:  30 seconds
.PP
However, for 1000 Mbit interfaces updating once every minute is usually a
working solution.
.PP
Estimated traffic values are likely to be somewhat inaccurate if daily
traffic is low because only the MB counter is used to calculate the
estimate.
.PP
Virtual and aliased interfaces can't be monitored because the kernel doesn't
provide traffic information for that type of interfaces. Such interfaces are
usually named eth0:0, eth0:1, eth0:2 etc. where eth0 is the actual interface
being aliased.
.SH AUTHOR
Teemu Toivola <tst at iki dot fi>, 
Wu Zhangjin <wuzhangjin at gmail dot com>
.SH "SEE ALSO"
.BR proc (5),
.BR ifconfig (8)


ChangeLog

	ChangeLog

	PLEASE add the latest change log on the top of THIS file, and
	follow this guidelines:
		[*] list the change log under the version number
		[*] list the change Time and your Email(P)
		[*] list the description of your change(D)

--------

[*] Version 1.0.0

3-Jul-2008

P: wu zhangjin, <wuzhangjin@gmail.com>
D: Arrange version 0.1.x and release a new version
D: Support multi-hosts via a light-weight http proxy
D: Reduce the SIZE of the XML output via using short XML flags
D: Substitude strftime() in gawk for the date command in shell version
D: Rearrange the XSL file, and write comments to every lines
D: Write vnStatXML to dump the data of vnStat in XML format
D: Use setTimeout/clearTimeout to monitor the Memory information
D: Add the configuration/Installation scripts

22-Oct-2008

P: wu zhangjin, <wuzhangjin@gmail.com>
D: add a tool vnstatgui to generate the graphical output in localhost.
D: it is src/vnstatgui

03-May-2009

P: wu zhangjin, <wuzhangjin@gmail.com>
D: bugfix: fix the parsing of addr:port 

28-Jul-2013

P: wu zhangjin, <wuzhangjin@gmail.com> of TinyLab, tinylab.org
D: bugfix: add busybox awk support
D: Add chromium browser support via fixing up vnstat.xsl and vnstat.js
D: Check httpd, httpd2, awk if no apache2 and gawk exist
D: Fix up compilier optimization of vnstatxml: -O2 -> -Os
D: Fix up linking issue, should use -lm after the other objectives
D: Add cross compiling support
D: Document the usage of local graphic output with vnstatgui
D: Verify and support Busybox based embedded system

28-Jul-2013

P: wu zhangjin, <wuzhangjin@gmail.com> of TinyLab, tinylab.org
D: Add multi protocols support, allows to collect data from remote hosts will any protocols
D: Allows to test the src/cgi-bin/vnstat-{c,p,shell}.sh scripts locally
D: Fix up the output format of the first column


src/cgi-bin/vnstatxml-1.6/src/vnstat.c
/*
vnStat - Copyright (c) 2002-08 Teemu Toivola <tst@iki.fi>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; version 2 dated June, 1991.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program;  if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave., Cambridge, MA 02139, USA.
*/

#include "vnstat.h"
#include "ifinfo.h"
#include "db.h"
#include "dbaccess.h"
#include "misc.h"
#include "cfg.h"

int main(int argc, char *argv[]) {

	int i;
	int currentarg, update=0, query=1, newdb=0, reset=0, sync=0;
	int active=-1, files=0, force=0, cleartop=0, rebuildtotal=0, traffic=0;
	int livetraffic=0;
	char interface[32], dirname[512], nick[32];
	char definterface[32], cfgfile[512];
	time_t current;
	DIR *dir;
	struct dirent *di;

	debug = 0; /* debug disabled by default */
	cfgfile[0] = '\0';

	/* early check for debug and config parameter */
	if (argc > 1) {
		for (currentarg=1; currentarg<argc; currentarg++) {
			if ((strcmp(argv[currentarg],"-D")==0) || (strcmp(argv[currentarg],"--debug")==0)) {
				debug = 1;
			} else if (strcmp(argv[currentarg],"--config")==0) {
				if (currentarg+1<argc) {
					strncpy(cfgfile, argv[currentarg+1], 512);
					if (debug)
						printf("Used config file: %s\n", cfgfile);
					currentarg++;
					continue;
				} else {
					printf("Error:\nFile for --config missing.\n");
					return 1;
				}
			}
		}
	}
	
	/* load config if available */
	loadcfg(cfgfile);

	/* set the default configuration for vnStatXML*/
	defaultcfg4vnstatxml();

	setlocale(LC_ALL, cfg.locale);
	strncpy(interface, "default", 32);
	strncpy(definterface, cfg.iface, 32);
	strncpy(nick, "none", 32);

	current = time(NULL);

	/* init dirname */
	strncpy(dirname, cfg.dbdir, 512);

	/* parse parameters, maybe not the best way but... */
	for (currentarg=1; currentarg<argc; currentarg++) {
		if (debug)
			printf("arg %d: \"%s\"\n",currentarg,argv[currentarg]);
		if (strcmp(argv[currentarg],"--longhelp")==0) {
		
			printf(" vnStat %s by Teemu Toivola <tst at iki dot fi>\n\n", VNSTATVERSION);
			
			printf("   Update:\n");
			printf("         -u, --update          update database\n");
			printf("         -r, --reset           reset interface counters\n");
			printf("         --sync                sync interface counters\n");
			printf("         --enable              enable interface\n");
			printf("         --disable             disable interface\n");
			printf("         --nick                set a nickname for interface\n");
			printf("         --cleartop            clear the top10\n");
			printf("         --rebuildtotal        rebuild total transfers from months\n");
			
			printf("   Query:\n");
			printf("         -q, --query           query database\n");
			printf("         -h, --hours           show hours\n");
			printf("         -d, --days            show days\n");
			printf("         -m, --months          show months\n");
			printf("         -w, --weeks           show weeks\n");
			printf("         -t, --top10           show top10\n");
			printf("         -s, --short           use short output\n");
			printf("         --dumpdb              show database in parseable format\n");

			/* Added by Wu Zhangjin at Sat Jun 28 04:27:03 CST 2008 */
			printf("         --dumpxml             show database in XML format\n");		
			printf("                                                                   \n");
			printf("	 This option is new for dumping the summary,day,hour,month,\n");
			printf("	 top10 data info in XML format, you can use it like this:  \n");
			printf("                                                                   \n");
			printf("         $ vnstat --dumpxml -i IFACE [no option]|-d|-h|-m|-t       \n");		
			
			printf("   Misc:\n");
			printf("         -i,  --iface          select interface (default: %s)\n", definterface);
			printf("         -?,  --help           short help\n");
			printf("         -D,  --debug          show some additional debug information\n");
			printf("         -v,  --version        show version\n");
			printf("         -tr, --traffic        calculate traffic\n");
			printf("         -l,  --live           show transfer rate in real time\n");
			printf("         --config              select used config file\n");
			printf("         --showconfig          dump config file with current settings\n");
			printf("         --testkernel          check if the kernel is broken\n");
			printf("         --longhelp            display this help\n\n");

			printf("See also \"man vnstat\".\n");

			return 0;

		} else if ((strcmp(argv[currentarg],"-?")==0) || (strcmp(argv[currentarg],"--help")==0)) {

			printf(" vnStat %s by Teemu Toivola <tst at iki dot fi>\n\n", VNSTATVERSION);

			printf("         -q,  --query          query database\n");
			printf("         -h,  --hours          show hours\n");
			printf("         -d,  --days           show days\n");
			printf("         -m,  --months         show months\n");
			printf("         -w,  --weeks          show weeks\n");
			printf("         -t,  --top10          show top10\n");
			printf("         -s,  --short          use short output\n");
			printf("         -u,  --update         update database\n");			
			printf("         -i,  --iface          select interface (default: %s)\n", definterface);
			printf("         -?,  --help           short help\n");
			printf("         -v,  --version        show version\n");
			printf("         -tr, --traffic        calculate traffic\n");
			printf("         -l,  --live           show transfer rate in real time\n\n");

			printf("See also \"--longhelp\" for complete options list and \"man vnstat\".\n");

			return 0;

		} else if ((strcmp(argv[currentarg],"-i")==0) || (strcmp(argv[currentarg],"--iface")==0)) {
			if (currentarg+1<argc) {
				strncpy(interface, argv[currentarg+1], 32);
				if (debug)
					printf("Used interface: %s\n", interface);
				currentarg++;
				continue;
			} else {
				printf("Error:\nInterface for -i missing.\n");
				return 1;
			}
		} else if (strcmp(argv[currentarg],"--config")==0) {
			/* config has already been parsed earlier so not but to do here*/
			currentarg++;
			continue;
		} else if ((strcmp(argv[currentarg],"--nick"))==0) {
			if (currentarg+1<argc) {
				strncpy(nick, argv[currentarg+1], 32);
				if (debug)
					printf("Used nick: %s\n", nick);
				currentarg++;
				continue;
			} else {
				printf("Error:\nNick for --nick missing.\n");
				return 1;
			}
		} else if ((strcmp(argv[currentarg],"-u")==0) || (strcmp(argv[currentarg],"--update")==0)) {
			update=1;
			query=0;
			if (debug)
				printf("Updating database...\n");
		} else if ((strcmp(argv[currentarg],"-q")==0) || (strcmp(argv[currentarg],"--query")==0)) {
			query=1;
		} else if ((strcmp(argv[currentarg],"-D")==0) || (strcmp(argv[currentarg],"--debug")==0)) {
			debug=1;
		} else if ((strcmp(argv[currentarg],"-d")==0) || (strcmp(argv[currentarg],"--days")==0)) {
			cfg.qmode=1;
		} else if ((strcmp(argv[currentarg],"-m")==0) || (strcmp(argv[currentarg],"--months")==0)) {
			cfg.qmode=2;
		} else if ((strcmp(argv[currentarg],"-t")==0) || (strcmp(argv[currentarg],"--top10")==0)) {
			cfg.qmode=3;
		} else if ((strcmp(argv[currentarg],"-s")==0) || (strcmp(argv[currentarg],"--short")==0)) {
			cfg.qmode=5;
		} else if ((strcmp(argv[currentarg],"-w")==0) || (strcmp(argv[currentarg],"--weeks")==0)) {
			cfg.qmode=6;			
		} else if ((strcmp(argv[currentarg],"-h")==0) || (strcmp(argv[currentarg],"--hours")==0)) {
			cfg.qmode=7;
		} else if (strcmp(argv[currentarg],"--dumpdb")==0) {
			cfg.qmode=4;
			
		/* Added by Wu Zhangjin at Sat Jun 28 04:12:56 CST 2008 */
		} else if (strcmp(argv[currentarg],"--dumpxml")==0) {
			cfg.xml.dumpxml=1;

		} else if (strcmp(argv[currentarg],"--enable")==0) {
			active=1;
			query=0;
		} else if ((strcmp(argv[currentarg],"-tr")==0) || (strcmp(argv[currentarg],"--traffic")==0)) {
			if (currentarg+1<argc) {
				if (isdigit(argv[currentarg+1][0])) {
					cfg.sampletime=atoi(argv[currentarg+1]);
					currentarg++;
					traffic=1;
					query=0;
					continue;
				}
			}
			traffic=1;
			query=0;
		} else if ((strcmp(argv[currentarg],"-l")==0) || (strcmp(argv[currentarg],"--live")==0)) {
			livetraffic=1;
			query=0;
		} else if (strcmp(argv[currentarg],"--force")==0) {
			force=1;
		} else if (strcmp(argv[currentarg],"--cleartop")==0) {
			cleartop=1;
		} else if (strcmp(argv[currentarg],"--rebuildtotal")==0) {
			rebuildtotal=1;
		} else if (strcmp(argv[currentarg],"--disable")==0) {
			active=0;
			query=0;
		} else if (strcmp(argv[currentarg],"--testkernel")==0) {
			i=kerneltest();
			return i;
		} else if (strcmp(argv[currentarg],"--showconfig")==0) {
			printcfgfile();
			return 0;			
		} else if ((strcmp(argv[currentarg],"-v")==0) || (strcmp(argv[currentarg],"--version")==0)) {
			printf("vnStat %s by Teemu Toivola <tst at iki dot fi>\n", VNSTATVERSION);
			return 0;
		} else if ((strcmp(argv[currentarg],"-r")==0) || (strcmp(argv[currentarg],"--reset")==0)) {
			reset=1;
			query=0;
		} else if (strcmp(argv[currentarg],"--sync")==0) {
			sync=1;
			query=0;			
		} else {
			printf("Unknown arg \"%s\". Use --help for help.\n",argv[currentarg]);
			return 1;
		}
		
	}

	/* check if the database dir exists and if it contains files */
	if (!traffic && !livetraffic) {
		if ((dir=opendir(dirname))!=NULL) {
			if (debug)
				printf("Dir OK\n");
			while ((di=readdir(dir))) {
				if (di->d_name[0]!='.') {
					strncpy(definterface, di->d_name, 32);
					files++;
				}
			}
			if (debug)
				printf("%d file(s) found\n", files);
			if (files>1) {
				strncpy(definterface, cfg.iface, 32);
			}
			closedir(dir);
		} else {
			printf("Error:\nUnable to open database directory \"%s\".\n", dirname);
			printf("Make sure it exists and is at least read enabled for current user.\n");
			printf("Exiting...\n");
			return 1;
		}
	}

	/* counter reset */
	if (reset) {
		if (!spacecheck(dirname) && !force) {
			printf("Error:\nNot enough free diskspace available.\n");
			return 1;
		}
		if (strcmp(interface, "default")==0)
			strncpy(interface, definterface, 32);
		readdb(interface, dirname);
		data.currx=0;
		data.curtx=0;
		writedb(interface, dirname, 0);
		if (debug)
			printf("Counters reseted for \"%s\"\n", data.interface);
	}

	/* counter sync */
	if (sync) {
		if (!spacecheck(dirname) && !force) {
			printf("Error:\nNot enough free diskspace available.\n");
			return 1;
		}
		if (strcmp(interface, "default")==0)
			strncpy(interface, definterface, 32);
		synccounters(interface, dirname);
		if (debug)
			printf("Counters synced for \"%s\"\n", data.interface);
	}

	/* clear top10 */
	if (cleartop) {
		if (!spacecheck(dirname) && !force) {
			printf("Error:\nNot enough free diskspace available.\n");
			return 1;
		}
		if (strcmp(interface, "default")==0)
			strncpy(interface, definterface, 32);
		if (force) {
			readdb(interface, dirname);

			for (i=0; i<=9; i++) {
				data.top10[i].rx=data.top10[i].tx=0;
				data.top10[i].rxk=data.top10[i].txk=0;
				data.top10[i].used=0;
			}

			writedb(interface, dirname, 0);
			printf("Top10 cleared for interface \"%s\".\n", data.interface);
			query=0;
		} else {
			printf("Warning:\nThe current option would clear the top10 for \"%s\".\n", interface);
			printf("Use --force to override this message.\n");
			return 1;
		}
	}

	/* rebuild total */
	if (rebuildtotal) {
		if (!spacecheck(dirname)) {
			printf("Error:\nNot enough free diskspace available.\n");
			return 1;
		}
		if (strcmp(interface, "default")==0)
			strncpy(interface, definterface, 32);
		if (force) {
			readdb(interface, dirname);

			data.totalrx=data.totaltx=data.totalrxk=data.totaltxk=0;
			for (i=0; i<=29; i++) {
				if (data.month[i].used) {
					addtraffic(&data.totalrx, &data.totalrxk, data.month[i].rx, data.month[i].rxk);
					addtraffic(&data.totaltx, &data.totaltxk, data.month[i].tx, data.month[i].txk);
				}
			}

			writedb(interface, dirname, 0);
			printf("Total transfer rebuild completed for interface \"%s\".\n", data.interface);
			query=0;
		} else {
			printf("Warning:\nThe current option would rebuild total tranfers for \"%s\".\n", interface);
			printf("Use --force to override this message.\n");
			return 1;
		}
	}

	/* enable & disable */
	if (active==1) {
		if (!spacecheck(dirname) && !force) {
			printf("Error:\nNot enough free diskspace available.\n");
			return 1;
		}
		if (strcmp(interface, "default")==0)
			strncpy(interface, definterface, 32);
		newdb=readdb(interface, dirname);
		if (!data.active && !newdb) {
			data.active=1;
			writedb(interface, dirname, 0);
			if (debug)
				printf("Interface \"%s\" enabled.\n", data.interface);
		} else if (!newdb) {
			printf("Interface \"%s\" is already enabled.\n", data.interface);
		}
	} else if (active==0) {
		if (!spacecheck(dirname) && !force) {
			printf("Error:\nNot enough free diskspace available.\n");
			return 1;
		}
		if (strcmp(interface, "default")==0)
			strncpy(interface, definterface, 32);
		newdb=readdb(interface, dirname);
		if (data.active && !newdb) {
			data.active=0;
			writedb(interface, dirname, 0);
			if (debug)
				printf("Interface \"%s\" disabled.\n", data.interface);
		} else if (!newdb) {
			printf("Interface \"%s\" is already disabled.\n", data.interface);
		}	
	}
	
	/* update */
	if (update) {
	
		/* check that there's some free diskspace left */
		if (!spacecheck(dirname) && !force) {
			printf("Error:\nNot enough free diskspace available.\n");
			return 1;
		}
		
		/* update every file if -i isn't specified */
		if (strcmp(interface, "default")==0) {
			
			dir=opendir(dirname);

			files=0;
			while ((di=readdir(dir))) {
				if (di->d_name[0]!='.') {
					files++;
					strncpy(interface, di->d_name, 32);
					if (debug)
						printf("\nProcessing file \"%s/%s\"...\n", dirname, interface);
					newdb=readdb(interface, dirname);
					if (data.active) {
						getifinfo(data.interface);
						parseifinfo(newdb);
						
						/* check that the time is correct */
						if ((current>=data.lastupdated) || force) {
							writedb(interface, dirname, newdb);
						} else {
							printf("Error:\nThe previous update was after the current date.\n\n");
							printf("Previous update: %s", (char*)asctime(localtime(&data.lastupdated)));
							printf("   Current time: %s\n", (char*)asctime(localtime(&current)));
							printf("Use --force to override this message.\n");
							return 1;
						}
					} else {
						if (debug)
							printf("Disabled interface \"%s\" not updated.\n", data.interface);
					}				
				}
			}

			closedir(dir);
			if (files==0) {
				// printf("No database found.\n");
				update=0;
			}
			
			/* reset to default */
			strncpy(interface, "default", 32);
		
		/* update only selected file */
		} else {
			newdb=readdb(interface, dirname);
			if (data.active) {
				getifinfo(data.interface);
				parseifinfo(newdb);
				if ((current>=data.lastupdated) || force) {
					if (strcmp(nick, "none")!=0)
						strncpy(data.nick, nick, 32);
					writedb(interface, dirname, newdb);
				} else {
					printf("Error:\nThe previous update was after the current date.\n\n");
					printf("Previous update: %s", (char*)asctime(localtime(&data.lastupdated)));
					printf("   Current time: %s\n", (char*)asctime(localtime(&current)));
					printf("Use --force to override this message.\n");
					return 1;
				}
			} else {
				if (debug)
					printf("Disabled interface \"%s\" not updated.\n", data.interface);
			}
		}
	}
	
	/* show databases */
	if (query) {
	
		/* show all interfaces if -i isn't specified */
		if (strcmp(interface, "default")==0) {
			
			if (files==0) {
				// printf("No database found.\n");
				query=0;
			} else if ((cfg.qmode==0) && (files>1)) {

				dir=opendir(dirname);
				printf("\n                     rx      /     tx      /    total    /  estimated\n");
				while ((di=readdir(dir))) {
					if (di->d_name[0]!='.') {
						strncpy(interface, di->d_name, 32);
						if (debug)
							printf("\nProcessing file \"%s/%s\"...\n", dirname, interface);
						readdb(interface, dirname);
						if (!newdb)
							showdb(5);
						
					}
				}
				closedir(dir);
				
			/* show in qmode if there's only one file or qmode!=0 */
			} else {
				readdb(definterface, dirname);
				if (!newdb) {
					if (cfg.qmode==5)
						printf("\n                     rx      /     tx      /    total    /  estimated\n");
					showdb(cfg.qmode);

				}
			}
		
		/* show only specified file */
		} else {
			readdb(interface, dirname);
			if (!newdb) {
				if (! cfg.xml.dumpxml) {
					if (cfg.qmode==5)
						printf("\n                     rx      /     tx      /    total    /  estimated\n");
					showdb(cfg.qmode);
				} else {
					dumpXML(cfg.qmode);
				}
			}
		}
	}
	
	/* calculate traffic */
	if (traffic) {
		if (strcmp(interface, "default")==0)
			strncpy(interface, definterface, 32);
		trafficmeter(interface, cfg.sampletime);
	}

	/* live traffic */
	if (livetraffic) {
		if (strcmp(interface, "default")==0)
			strncpy(interface, definterface, 32);

		livetrafficmeter(interface);
	}
	
	/* if nothing was shown previously */
	if (!query && !update && !reset && !sync && active==-1 && !cleartop && !rebuildtotal && !traffic && !livetraffic) {
		
		/* give more help if there's no database */
		if (files==0) {
			printf("No database found, nothing to do. Use --help for help.\n\n");
			printf("A new database can be created with the following command:\n");
			printf("    %s -u -i eth0\n\n", argv[0]);
			printf("Replace 'eth0' with the interface that should be monitored. A list\n");
			printf("of available interfaces can be seen with the 'ifconfig' command.\n");
		} else {
			printf("Nothing to do. Use --help for help.\n");
		}
	}
	
	return 0;
}
